docs/source/package-spec.rst

```diff
@@ -746,6 +778,18 @@ to the `Compiler Input and Output Description <http://solidity.readthedocs.io/en
   :Key: ``settings``
   :Type: Object
 
+Contract Types: ``contractTypes``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+A list of the |ContractAlias| in this package that used this compiler to generate its outputs. 
+
+  - All ``contractTypes`` that locally declare ``runtimeBytecode`` **should** be attributed for by a compiler object.
+  - A single ``contractTypes`` **must** not be attributed to more than one compiler.
```

Comment:
I think I see here is why the compiler bit is an array yes?

Reply:
@pipermerriam Yeah, the general conclusion was that the `compiler` info is better suited as an input, when the rest of `contractTypes` info is more of an output (`abi`, `bytecodes`, etc), so it would be cleaner to separate the two. Especially when it comes to using packages as compiler json inputs - considering also packages with `.vy` & `.sol` sources - a compiler doesn't have to parse all the sources to figure out the relevant files.

There are also some DRY benefits - eg. when many `contractTypes` share the same compiler.

One question I have is - along the lines of DRY. I'm trying to think if it's reasonable to drop `compiler` / `runtime_bytecode` from a `deployment` - and if a user wants to do any validation require that a corresponding `contractType` is available. But, this could also just result in an excessive amount of duplication in `contractTypes` to account for many deployments w/ different `linkReferences`.

Reply:
I'm sorry but I don't follow your question.  Can you pose it with a simple example maybe of the before/after?

Reply:
Sure! Essentially I'm trying to figure out if we can remove the `"compiler"` object from  a `"deployments"` object.

v2: Compilers can be defined within a deployment. The spec requires that a `"deployment"` defines a `"contract_type"` and that the specified `"contract_type"` (eg `"Escrow"`) is defined within the package's top-level `"contract_types"` - which also can contain a `"compiler"` object
```json
{
  "contract_types": {
    "Escrow": {
      "compiler": {...}
    }
  },
  "deployments": {
    "blockchain://abc/block/123": {
     "Escrow": {
        "contract_type": "Escrow",
        "address": "0x8c1...",
        "transaction": "0xc60...",
        "block": "0x4d1a...",
        "runtime_bytecode": {
          "link_dependencies": [
            {"offsets": [301, 495], "type": "reference", "value": "SafeSendLib"}
          ]
        },
        "compiler": {
          "name": "solc",
          "version": "0.4.24+commit.e67f0147.Emscripten.clang",
          "settings": {
            "optimize": true
          }
        }
      }
  }
}
```

v3: Moving compiler information to a top-level `"compilers"`, this should be sufficient to match a compiler to all `"contractTypes"` - and their respective `"deployments"`. Allowing us to remove the `"compiler"` object from within a `"deployments"` object.
```json
{
  "compilers": [{
    "name": "solc",
    "version": "0.4.24....",
    "contractTypes": ["Escrow"],
  }]
}
```

Let me know if that helps - or more clarification is needed.

Reply:
:+1: